### PR TITLE
[MI-3442] Optimise subscriptions monitor system

### PR DESCRIPTION
**Pull request by ayusht2810**
#### Summary

- Optimise subscriptions monitor system.
- Remove refreshing of subscriptions during processing of activity.

### CODE REVIEW

File: server/msteams/client.go [hunk 1]
```diff
@@ -848,6 +849,51 @@ func (tc *ClientImpl) DeleteSubscription(subscriptionID string) error {
...
+		return nil, NormalizeGraphAPIError(err)
+	}
+
+	pageIterator, err := msgraphcore.NewPageIterator[models.Subscriptionable](r, tc.client.GetAdapter(), models.CreateSubscriptionCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return nil, NormalizeGraphAPIError(err)
+	}
+
+	subscriptions := []*Subscription{}
+	err = pageIterator.Iterate(tc.ctx, func(u models.Subscriptionable) bool {
```
Comment (raghavaggarwal2308) [inline server/msteams/client.go:864]: What does `u` specify here. Use proper name if possble

File: server/plugin.go [hunk 2]
```diff
@@ -221,81 +226,155 @@ func (p *Plugin) startSubscriptions() {
...
-				SubscriptionID: "fake-subscription-id",
-				Type:           "allChats",
-				ExpiresOn:      time.Now(),
-				Secret:         p.getConfiguration().WebhookSecret,
-			})
-			<-ws
-			return
-		}
-		p.API.LogDebug("Subscription to all chats created", "subscriptionID", chatsSubscription.ID)
+		// Check if all chats subscription is present on Mattermost
```
Comment (raghavaggarwal2308) [inline server/plugin.go:259]:
```suggestion
		// Check if all chat subscriptions are present on Mattermost
```
I think this is better. If changing then change the variable names also

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin.go (hunk 2 above)
Comment (ayusht2810) [reply server/plugin.go:259]: we are creating a single subscription for all the chats whose type is `allChats`. So, I think the current comment is fine already.

File: server/store/store.go [hunk 3]
```diff
@@ -53,9 +53,12 @@ type Store interface {
 	TeamsToMattermostUserID(userID string) (string, error)
 	MattermostToTeamsUserID(userID string) (string, error)
 	CheckEnabledTeamByTeamID(teamID string) bool
-	ListGlobalSubscriptionsToCheck() ([]storemodels.GlobalSubscription, error)
+	ListGlobalSubscriptions() ([]storemodels.GlobalSubscription, error)
```
Comment (manojmalik20) [inline server/store/store.go:56]:
```suggestion
	ListGlobalSubscriptions() ([]*storemodels.GlobalSubscription, error)
```
Why are we not using pointers in all these functions?

File: server/msteams/client.go [hunk 4]
```diff
@@ -848,6 +849,51 @@ func (tc *ClientImpl) DeleteSubscription(subscriptionID string) error {
...
+	pageIterator, err := msgraphcore.NewPageIterator[models.Subscriptionable](r, tc.client.GetAdapter(), models.CreateSubscriptionCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return nil, NormalizeGraphAPIError(err)
+	}
+
+	subscriptions := []*Subscription{}
+	err = pageIterator.Iterate(tc.ctx, func(subscription models.Subscriptionable) bool {
+		subscriptionID := ""
+		var expiresOn time.Time
+		resource := ""
```
Comment (manojmalik20) [inline server/msteams/client.go:867]:
```suggestion
		subscriptionID := ""
		resource := ""
        var expiresOn time.Time
```

File: server/store/storemodels/storemodels.go [hunk 5]
```diff
@@ -23,6 +23,7 @@ type GlobalSubscription struct {
 	SubscriptionID string
 	Type           string
 	ExpiresOn      time.Time
+	ExpiresOnEpoch int64
```
Comment (manojmalik20) [inline server/store/storemodels/storemodels.go:26]: What's the need for this?

File: server/monitor/subscriptions_test.go [hunk 6]
```diff
@@ -16,78 +16,108 @@ import (
...
 	for _, testCase := range []struct {
-		description string
-		setupClient func(*mocksClient.Client)
-		setupAPI    func(*plugintest.API)
-		setupStore  func(*mocksStore.Store)
+		description          string
+		setupClient          func(*mocksClient.Client)
+		setupAPI             func(*plugintest.API)
+		setupStore           func(*mocksStore.Store)
+		msteamsSubscription  map[string]*msteams.Subscription
```
Comment (manojmalik20) [inline server/monitor/subscriptions_test.go:29]:
```suggestion
		msteamsSubscriptionMap  map[string]*msteams.Subscription
```

File: server/monitor/subscriptions_test.go [hunk 7]
```diff
@@ -16,78 +16,108 @@ import (
...
+		allChatsSubscription *msteams.Subscription
 	}{
 		{
 			description: "Fail to get global subscription list",
 			setupClient: func(client *mocksClient.Client) {},
 			setupAPI: func(mockAPI *plugintest.API) {
 				mockAPI.On("LogDebug", "Checking for global subscriptions").Times(1)
-				mockAPI.On("LogError", "Unable to get the global subscriptions", "error", mock.Anything).Times(1)
+				mockAPI.On("LogError", "Unable to get the chat subscriptions from store", "error", mock.Anything).Times(1)
```
Comment (manojmalik20) [inline server/monitor/subscriptions_test.go:37]: Remove the use of `mock.Anything` and replace `err` with `err.Error()`

File: server/monitor/subscriptions_test.go [hunk 8]
```diff
@@ -135,147 +170,77 @@ func TestMonitorCheckChannelSubscriptions(t *testing.T) {
...
+				client.On("RefreshSubscription", "test").Return(nil, errors.New("unable to refresh the subscription"))
 				client.On("DeleteSubscription", "test").Return(nil)
-				client.On("SubscribeToUserChats", "user-id", "base-url", "webhook-secret", true).Return(&msteams.Subscription{ID: "new-id", ExpiresOn: newExpiresOn}, nil)
+				client.On("SubscribeToChannel", "team-id", "channel-id", "base-url", "webhook-secret").Return(&msteams.Subscription{ID: "new-id", ExpiresOn: newExpiresOn}, nil)
 			},
 			setupAPI: func(mockAPI *plugintest.API) {
-				mockAPI.On("LogDebug", "Checking for chats subscriptions").Times(1)
-				mockAPI.On("LogDebug", "Refreshing chats subscriptions", "count", 1).Times(1)
+				mockAPI.On("LogDebug", "Checking for channels subscriptions")
+				mockAPI.On("LogDebug", "Refreshing channels subscriptions", "count", 1)
```
Comment (manojmalik20) [inline server/monitor/subscriptions_test.go:219]: Why removed `.Times(1)`?

File: server/monitor/monitor.go [hunk 9]
```diff
@@ -87,29 +86,12 @@ func (m *Monitor) Stop() {
 func (m *Monitor) check() {
 	msteamsSubscriptionsMap, allChatsSubscription, err := m.GetMSTeamsSubscriptionsMap()
 	if err != nil {
-		m.api.LogError("Unable to fetch subscriptions from MS Teams", "error", err)
+		m.api.LogError("Unable to fetch subscriptions from MS Teams", "error", err.Error())
 		return
 	}
 
 	m.checkGlobalSubscriptions(msteamsSubscriptionsMap, allChatsSubscription)
```
Comment (manojmalik20) [inline server/monitor/monitor.go:93]: Why not put this in goroutine? How will these run in parallel this way?

**Review by manojmalik20 — COMMENTED** [review]
Please fix the failing CI.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**